pkg/scheduler/objects/application.go

```diff
@@ -225,6 +234,56 @@ func (sa *Application) clearStateTimer() {
 		zap.String("state", sa.stateMachine.Current()))
 }
 
+func (sa *Application) isWaitingStateTimedOut() bool {
+	return sa.IsWaiting() && sa.stateTimer == nil
+}
+func (sa *Application) initPlaceholderTimer() {
+	if sa.placeholderTimer != nil || !sa.IsAccepted() || sa.execTimeout <= 0 {
+		return
+	}
+	log.Logger().Debug("Application placeholder timer initiated",
+		zap.String("AppID", sa.ApplicationID),
+		zap.Duration("Timeout", sa.execTimeout))
+	sa.placeholderTimer = time.AfterFunc(sa.execTimeout, sa.timeoutPlaceholderProcessing)
+}
+
+func (sa *Application) clearPlaceholderTimer() {
+	if sa == nil || sa.placeholderTimer == nil {
+		return
+	}
+	sa.placeholderTimer.Stop()
+	sa.placeholderTimer = nil
+}
+
+func (sa *Application) timeoutPlaceholderProcessing() {
```

Comment:
Case 1: The app has made at least one or more replacements and is really doing things. It could take days or weeks for an app to transition to completed. You do not want to keep unused placeholders using resources for that long so for case 1 you clean up unused allocated placeholder. You do not want to touch the app itself as it runs. There are no pending placeholder asks anymore (impossible).

Case 2: is the case that one or more placeholders have been allocated but not all have been and the app is still waiting for the left over placeholders to be allocated. No real allocations have been made. You remove all already allocated placeholders and the left over asks. We also fail the application to show something went wrong.